buckets/local/bucket.go

```diff
@@ -0,0 +1,297 @@
+	var links []*ipld.Link
+	for _, l := range n.Links() {
+		if l.Name != "" {
+			links = append(links, l)
+		}
+	}
+	return links, nil
+}
+
+// walkPath walks path and adds files to the dag service.
+func (b *Bucket) walkPath(ctx context.Context, pth string) (ipld.Node, error) {
+	root := unixfs.EmptyDirNode()
+	prefix, err := md.PrefixForCidVersion(1)
+	if err != nil {
+		return nil, err
+	}
+	root.SetCidBuilder(prefix)
+	editor := dagutils.NewDagEditor(root, b.tmp)
+	abs, err := filepath.Abs(pth)
+	if err != nil {
+		return nil, err
+	}
+	if err = filepath.Walk(abs, func(n string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			if Ignore(n) {
+				return nil
+			}
+			p := n
+			n = strings.TrimPrefix(n, abs+"/")
+			if strings.HasPrefix(n, filepath.Dir(archiveName)+"/") || strings.HasSuffix(n, PatchExt) {
+				return nil
+			}
+			file, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+			nd, err := addFile(b.tmp, b.layout, file)
+			if err != nil {
+				return err
+			}
+			if err = editor.InsertNodeAtPath(ctx, n, nd, unixfs.EmptyDirNode); err != nil {
+				return err
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return editor.Finalize(ctx, b.tmp)
+}
+
+// createArchive creates the empty archive file under path.
+func (b *Bucket) createArchive() (*os.File, error) {
+	name := filepath.Join(b.path, archiveName)
+	if err := os.MkdirAll(filepath.Dir(name), os.ModePerm); err != nil {
+		return nil, err
+	}
+	return os.Create(name)
+}
+
+// ArchiveFile creates an archive describing a directory containing reader.
+func (b *Bucket) ArchiveFile(ctx context.Context, pth string, name string) error {
+	r, err := os.Open(pth)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	root := unixfs.EmptyDirNode()
+	prefix, err := md.PrefixForCidVersion(1)
+	if err != nil {
+		return err
+	}
+	root.SetCidBuilder(prefix)
+	editor := dagutils.NewDagEditor(root, b.tmp)
+	f, err := addFile(b.tmp, b.layout, r)
+	if err != nil {
+		return err
+	}
+	if err = editor.InsertNodeAtPath(ctx, name, f, unixfs.EmptyDirNode); err != nil {
+		return err
+	}
+	n, err := editor.Finalize(ctx, b.tmp)
+	if err != nil {
+		return err
+	}
+	file, err := b.createArchive()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if err = car.WriteCarWithWalker(ctx, b.tmp, []cid.Cid{n.Cid()}, file, carWalker); err != nil {
+		return err
+	}
+	b.root = n.Cid()
+	return nil
+}
+
+// HashFile returns the cid of the file at path.
+func (b *Bucket) HashFile(pth string) (cid.Cid, error) {
+	r, err := os.Open(pth)
+	if err != nil {
+		return cid.Undef, err
+	}
+	defer r.Close()
+	n, err := addFile(b.tmp, b.layout, r)
+	if err != nil {
+		return cid.Undef, err
+	}
+	return n.Cid(), nil
+}
+
+// Diff returns a list of changes that are present in pth compared to the current archive.
+func (b *Bucket) Diff(ctx context.Context, pth string) (diff []*dagutils.Change, err error) {
```

Comment:
Generates a diff between the current archive and a directory... used to create a change list.